### DISCUSSION TITLE
Juce or not to Juce...

### DISCUSSION THREAD

**Discussion by Ted-juh** [RomanKubiak/ctrlr#439] [General]
Perhaps it's like swearing/ cursing in a church but how many of you guys tried to make a panel in Juce instead of CTRLR? 

I started a few weeks ago and didn't get much further than a few sliders with some L&F applied to it.  The fact that I'm a noob concerning C++ doesn't help either. C++ is an instant headache. Take into account the whole structure of Juce with its overrides/ namespaces, do this here and that over there in another file without any explanation why (and do not dare to ask simple questions on the forum) and it might be clear that Ctrlr is a breeze to work with. Although it has its limitations.

All tutorials seem to cover just the basics or are copying the tutorials of Juce itself which in itself are very confusing to me. I found some books on Juce/ C++ but the books about Juce are getting too old and I haven't found a book about C++ without a lot of foo bar or a/ b which confuses the hell out of me. 

Then why do I start with Juce and C++? Well, I thought if I could get a better understanding of Juce/ C++ I could help improve the CTRLR project. I'm not reluctant to put some effort into it but it's like hitting a brick wall every 10 minutes. So the question is, do you guys have any recommendations where to start to learn C++ or Juce? Or do you have good tips on how to take this hurdle?

**Comment by damiensellier** [comment]
Verrrry good question.
I thought about migrating too, for several reasons, but the amount of skills required to create just a basic panel is huge.
This is why Ctrlr is such a nice app because it's (almost) wysiwyg and you don't need the hassle of C++ programming and brainkilling that a JUCE project requires. I dived into JUCE during christmas time and I was having burnout nightmares of code etc at night lol for real.

Unless you want to process audio, I would prefer focusing on optimizing Ctrlr by adding JUCE classes and enhancing the workflow to focus on beautiful working panels. Ctrlr can really be improved, having all jUCE color schemes would be great for exemple, some JUCE class are not bound to LUA and so on... 


**The pros of all JUCE**
Smaller build size (not the +32Mo for a basic Ctrlr panel)
Full implementation of JUCE classes
Possibility to add C++ extra features and libraries


**The Cons of all JUCE**
Requires advanced/pro C++ skills and practices
Requires building all apps/plugins with xCode and VS2019
No GUI editor for easy panel design


For midi only interfaces I will forever stick to Ctrlr. Someday, if I have extra freetime I'll try to fix and add extra JUCE features to Ctrlr, but I am not a developper and definitely prefer doing music instead of coding :) So Ctrlr is the right way to make midi plugins to me.

**Reply by bijlevel** [reply]
Same here! I did succeed on building an AU plugin based on a tutorial. But I agree with @damiensellier , JUCE is more for audio processing. The biggest strength of CTRLR is the ability to do all sorts of things with MIDI sysex, which makes it indeed the number one candidate for replacing hardware interfaces with software! Maybe MAX/MSP comes near. It certainly is very user friendly and simple panels are a breeze to build, but I hit the wall when In had to process binary data. Support from the staff and the forum is real good!

**Comment by Ted-juh** [comment]
Have to agree with both of you about the pros and cons. It's the GUI-editor of Ctrlr that smashes Juce out of the ballpark. But in my humble opinion, as soon as there is LUA involved to go beyond what's capable with the GUI-editor, Ctrlr has a steep learning curve as well. I think the (old) forum proves my point here despite its helpful members.

In the meantime, I've been doing my research about JUCE. Found some tutorials that might help one another. Here's a small list:

http://scowlingowlsound.com/blog/2019/blog-1
If you are a true beginner, this is a must. (If you don't mind the sarcastic humor and cursing.)

https://www.youtube.com/channel/UCpKb02FsH4WH4X_2xhIoJ1A
The Audio Programmer. Need I say more? In my opinion, he copies the Juce tutorials a little too often with just different variables. Nonetheless, it's a good walkthrough of what it takes to produce something useful in JUCE. He also started a series about C++ which were helpful to me.

https://suzuki-kengo.dev/
Lots of tutorials for the GUI part of JUCE. Very helpful.
